src/java_bytecode/java_object_factory.cpp

```diff
@@ -68,6 +69,16 @@ class java_object_factoryt
   /// this set AND the tree depth becomes >= than the maximum value above.
   std::unordered_set<irep_idt, irep_id_hash> recursion_set;
 
+  /// Every time the non-det generator visits a type and the type is generic
+  /// (either a struct or a pointer), the following map is used to store and
+  /// look up the concrete types of the generic paramaters in the current
+  /// scope. Note that not all generic parameters need to have a concrete
+  /// type, e.g., the method under test is generic. The types are removed
+  /// from the map when the scope changes. Note that in different depths
+  /// of the scope the parameters can be specialized with different types
+  /// so we keep a stack of types for each parameter.
+  generic_recursion_mapt generic_recursion_map;
```

Comment:
Is `recursion` actually the right name? Sorry should have spotted earlier, but it is more of a `generic_type_specialisation_map` no?